### ISSUE TITLE
Why does the report shows solution that does not match the amount of thumbnails available ?

### ISSUE THREAD

**Issue by kopax** [ausi/respimagelint#47]
After testing my website, I have the following report for almost all my images:

![image](https://user-images.githubusercontent.com/1866564/103227137-f7855b00-492d-11eb-9f7c-cbccd8e1289b.png)

I do not understand how to fix this, the recommended settings is always: `Try using sizes="(min-width: 1000px) 896px, 92.35vw" instead.`, while I Have multiple thumbnails available.

- Is sizes limited to one small and one medium image? (two images)
- Should I use `min-width` or `max-width` for my media?
- I do not use `media` in `<source />` as I don't want to crope the image for small resolution, is that ok to just use the `srcSet` and `sizes`?

Thanks a lot for advising on how to get started with responsive image. I apologize in advance if this is not a place for asking questions.

**Comment by ausi** [comment]
> Is sizes limited to one small and one medium image? (two images)

No `sizes` is not limited to two sizes. The `sizes` attribute is not directly related to the number of images that are available via `srcset`. It tells the browser how large the image will be displayed (based on the CSS of the page) so that the browser has enough information to be able to pick the best candidate from `srcset`.

> Should I use `min-width` or `max-width` for my media?

Both are fine, I generally suggest to use the same form that is used in the stylesheet of the page.

> I do not use `media` in `<source />` as I don't want to crope the image for small resolution, is that ok to just use the `srcSet` and `sizes`?

Yes, this is exactly correct. `<source media=...>` should only be used for “art direction”.

> I apologize in advance if this is not a place for asking questions.

On large projects on GitHub I would recommend to look for other places (like forums or slack channels) to ask such kind of questions, so that the maintainers of those projects don’t get overloaded. On small projects (like this one) it’s usually OK to open a GitHub ticket for such topics I think 🙃

> I do not understand how to fix this, the recommended settings is always: `Try using sizes="(min-width: 1000px) 896px, 92.35vw" instead.`, while I Have multiple thumbnails available.

From the description above my guess would be that the linter is correct and that you should use the suggested `sizes` attribute.

I would recommend to take a look at <https://ericportis.com/posts/2014/srcset-sizes/> (especially “Part 2”) for a better understanding of how `sizes` and `srcset` work together.

This answer on Stack Overflow might be interesting too for your use case: <https://stackoverflow.com/a/39338181>

If you can put your testing website somewhere online and post the link to it here I can further investigate if the linter works correct in your case.

**Comment by kopax** [comment]
Hi @ausi and thanks for your deep reply, I appreciate it.

This is how I render the Image set using [`responsive-loader`](https://github.com/dazuaz/responsive-loader):

```jsx
import React, { useMemo } from 'react';
import Image from './Image';

export default function ImageSet({
  webp,
  src,
  srcSet,
  width,
  height,
  lazy = true,
  images,
  toString,
  ...rest
}) {
  const sizes = useMemo(() => {
    // return '(min-width: 1000px) 896px, 92.35vw';
    return images ?
      images.map((image, i) => i + 1 !== images.length ? `(max-width: ${image.width}px) ${image.width}px` : '100vw')
        .join(', ') :
      undefined;
  }, [images]);
  return (
    <picture>
      {webp && (
        <source
          type="image/webp"
          srcSet={webp.srcSet}
          sizes={sizes}
        />
      )}
      <source
        srcSet={srcSet}
        sizes={sizes}
      />
      <Image
        src={webp ? webp.src : src}
        sizes={sizes}
        loading={lazy ? 'lazy' : undefined}
        width={width}
        height={height}
        {...rest}
      />
    </picture>
  );
}
```

The props for this component look like follow:
```json
{
  "webp": {
    "srcSet": "/static/media/business-962358_1920-xl-492.396e1aaf.webp 492w,/static/media/business-962358_1920-xl-640.41098500.webp 640w,/static/media/business-962358_1920-xl-820.3f59a457.webp 820w,/static/media/business-962358_1920-xl-1200.e871831d.webp 1200w,/static/media/business-962358_1920-xl-1600.7daa281a.webp 1600w",
    "images": [
      {
        "path": "/static/media/business-962358_1920-xl-492.396e1aaf.webp",
        "width": 492,
        "height": 369
      },
      {
        "path": "/static/media/business-962358_1920-xl-640.41098500.webp",
        "width": 640,
        "height": 480
      },
      {
        "path": "/static/media/business-962358_1920-xl-820.3f59a457.webp",
        "width": 820,
        "height": 615
      },
      {
        "path": "/static/media/business-962358_1920-xl-1200.e871831d.webp",
        "width": 1200,
        "height": 900
      },
      {
        "path": "/static/media/business-962358_1920-xl-1600.7daa281a.webp",
        "width": 1600,
        "height": 1200
      }
    ],
    "src": "/static/media/business-962358_1920-xl-492.396e1aaf.webp",
    "placeholder": "data:image/webp;base64,UklGRqAAAABXRUJQVlA4IJQAAAAwBACdASoUAA8APlEejUQjoaEYBAA4BQS0gFtkxiJkzHwqlPueogMfmgAA/vhCvgxfzuHP+WgDnRbf0CTbl4We5hSQhks+r9D/sod106VD9TabYEZJxoApveAy/g1eP5Zm4W+B47QyefXXv6LxxhX5EArk2Atm+sDIDpnAfTl8BqCIIQbdokNtH4XCRAYksoAngAAA",
    "width": 492,
    "height": 369
  },
  "src": "/static/media/business-962358_1920-xl-492.6f79b25f.jpg",
  "srcSet": "/static/media/business-962358_1920-xl-492.6f79b25f.jpg 492w,/static/media/business-962358_1920-xl-640.4dba1f6c.jpg 640w,/static/media/business-962358_1920-xl-820.badc04c2.jpg 820w,/static/media/business-962358_1920-xl-1200.b5a125df.jpg 1200w,/static/media/business-962358_1920-xl-1600.acb97681.jpg 1600w",
  "width": 492,
  "height": 369,
  "lazy": true,
  "images": [
    {
      "path": "/static/media/business-962358_1920-xl-492.6f79b25f.jpg",
      "width": 492,
      "height": 369
    },
    {
      "path": "/static/media/business-962358_1920-xl-640.4dba1f6c.jpg",
      "width": 640,
      "height": 480
    },
    {
      "path": "/static/media/business-962358_1920-xl-820.badc04c2.jpg",
      "width": 820,
      "height": 615
    },
    {
      "path": "/static/media/business-962358_1920-xl-1200.b5a125df.jpg",
      "width": 1200,
      "height": 900
    },
    {
      "path": "/static/media/business-962358_1920-xl-1600.acb97681.jpg",
      "width": 1600,
      "height": 1200
    }
  ]
}
```

> Both are fine, I generally suggest to use the same form that is used in the stylesheet of the page.

I tried to replace `min-width` with `max-width` and none of the two helps to switch the image properly

> From the description above my guess would be that the linter is correct and that you should use the suggested `sizes` attribute.

I have not found a way to dynamically generate the `sizes` using those data, is that even possible? Why would I want to use the suggested `sizes` if I have more than 2 thumbnails available for more than just 2 viewport size?

> If you can put your testing website somewhere online and post the link to it here I can further investigate if the linter works correct in your case.


You can view the testing website [here](https://preprod.kopaxgroup.com)

Thanks in advance for helping

**Comment by ausi** [comment]
Taking the first image on your website: By resizing the browser window I can see that for viewports wider than 950px the image gets displayed at a width of exactly 896px. For all smaller viewports the image has a width of about 95% of the viewport. This is the information we put into the `sizes` attribute like this:

```html
sizes="(min-width: 950px) 896px, 95vw"
```

If you want to use `max-width` instead you can write it like so (this is functionally exactly equivalent):

```html
sizes="(max-width: 949px) 95vw, 896px"
```

If you combine this with your `srcset` attribute the browser is now able to select the correct image for your viewport/pixel density combination.

The final source code should look like this:

```html
<picture>
	<source 
		type="image/webp" 
		sizes="(min-width: 950px) 896px, 95vw"
		srcset="/static/media/architecture-1354811_1920-xl-492.55964624.webp 492w,/static/media/architecture-1354811_1920-xl-640.fcddd19e.webp 640w,/static/media/architecture-1354811_1920-xl-820.8c8b6443.webp 820w,/static/media/architecture-1354811_1920-xl-1200.acb47f16.webp 1200w,/static/media/architecture-1354811_1920-xl-1600.9a939562.webp 1600w"
	>
	<img
		sizes="(min-width: 950px) 896px, 95vw"
		srcset="/static/media/architecture-1354811_1920-xl-492.da0c4d98.jpg 492w,/static/media/architecture-1354811_1920-xl-640.0a58aabc.jpg 640w,/static/media/architecture-1354811_1920-xl-820.03ec185d.jpg 820w,/static/media/architecture-1354811_1920-xl-1200.06c1ce59.jpg 1200w,/static/media/architecture-1354811_1920-xl-1600.8961d518.jpg 1600w"
		src="/static/media/architecture-1354811_1920-xl-1600.8961d518.jpg"
		class="has-shadow"
		alt="Hero"
	>
</picture>
```

> I have just read the whole article and to be honest, I still do not understand how it will use more than 2 thumbnails using just two `sizes`.

The sizes attribute is not a set of multiple “sizes”. It tells the browser how wide the image will be shown on the page (based on markup and CSS). The `95vw` part of the `sizes` attribute in your example is dynamic and refers to the viewport with of the browser. So the quantity of the sizes in this attribute is not two in this case but more or less infinite.

> I have not found a way to dynamically generate the `sizes` using those data, is that even possible?

I don’t think it is because the displayed width of the image depends on the CSS code in combination with the HTML of the whole page. Dynamically generating the `sizes` attribute is very hard in most cases. Respimagelint does it by resizing the browser hundreds of times and measuring each image afterwards.

If your components are dynamic you might be able to set the `sizes` attribute to the rendered width of the parent component. But this makes your component depend on layout state and it would not work on server-side rendered components.

**Comment by kopax** [comment]
I think I got it all now. It's enriching answers, thanks a lot. I have removed the unnecessary `<source>` for media type jpeg, and I will use the tool to calculate the sizes for me, should this be good enough?

It would be awesome to have it done dynamically through webpack plugins and using a headless browser like puppeteer.
